scanner/decisions/0010-separate-versioning-for-vuln-bundles.md

```diff
@@ -0,0 +1,84 @@
+
+This approach needs to be revised for the following reasons:
+
+1.  The vulnerability schema rarely changes, but maintaining multiple bundles is expensive.  It requires multiple GitHub workflow runs, which increases the chances of failure and pressure on the security sources' endpoints.
+2.  It can be challenging to backport fixes and updates to previous releases.  This requires patch releases, which trigger the release workflow.  Improving the release process can be complex, as it may require patch release cuts, changes to different branches, and integration between logic that is managed in the master branch, executing other releases binaries.
+3.  The latest offline bundles does not support post `4.5` releases.  Bundle embedding for each release increases the cost to store and consume offline bundles.  The `roxctl` interface and documentation today needs to be clearer on these limitations.
+
+## Decision
+
+We will implement a versioning stream for vulnerability bundles, bumped when we determine that ClairCore's schema has changed.
+
+Initially, two version streams will exist (table 1 below):
+
+| Stream | Git Ref      | Description                                                        |
+|--------|--------------|--------------------------------------------------------------------|
+| dev    | heads/master | This stream tracks the latest vulnerability schema                 |
+| v1     | heads/4.4.5  | This stream tracks the schema used by the current StackRox release |
+
+On the subsequent releases (for example,. `4.4.6`), one of the two will be carried out:
+
+1.  The schema hasn't changed, so v1 now tracks the new release (`heads/4.4.6`).
+2.  The schema has changed, so v1 stays at the current release (`heads/4.4.5`), and a new version, v2, is created to track the next release (`heads/4.4.6`).
+
+### Building versioned bundles
+
+The GitHub workflow will have a map with "bundle version" as keys and "git reference" as values, similar to Table 1 above.  This can be stored statically in the matrix definition.
+
+For each version in the mapping, a bundle will be built based on the specified git ref and stored at `https://definitions.stackrox.io/v4/<bundle-version>/vulnerabilities.zip`.
+
+In every release cut, the key specified in `scanner/VULNERABILITY_VERSION` will be updated in the mapping to the current release branch reference (e.g., `heads/4.4.6`).  Notice that if the version is the same, it will be updated.  If it's new, it will be added, and the previous version will stay.
+
+Offline bundles per release will not be created anymore, but per bundle version.  The latest offline bundle will ship all bundle versions currently supported by StackRox.
```

Comment:
> The latest offline bundle will ship all bundle versions currently supported by StackRox.

What is the motivation for doing this? That would mean the latest offline bundle would contain (current state):

```
k8s-istio.zip
scanner-defs.zip
scanner-v4-defs-4.4.zip  <- new
scanner-v4-defs-4.5.zip
scanner-v4-defs-v1.zip   <- new
scanner-v4-defs-v2.zip   <- new 
etc.
```
Bundle would be approaching 1GB. This seems wasteful and cost prohibitive for an ACS install that would only ever need one version. We had previously communicated that there was no guarantee post `4.4` we would be adding additional versions to this bundle.

Reply:
> What is the motivation for doing this? That would mean the latest offline bundle would contain (current state):

Kinda, I think it would look like:

```
k8s-istio.zip
scanner-defs.zip
scanner-v4-defs-4.5.zip
scanner-v4-defs-v1.zip   <- new
```

Later, if we have v2 (while 4.5 is still supported) we would have:

```
k8s-istio.zip
scanner-defs.zip
scanner-v4-defs-4.5.zip
scanner-v4-defs-v1.zip
scanner-v4-defs-v2.zip   <- new
```

And once 4.5 is out, we can remove it from the bundle.

If the size is a concern, we could fashion the possibility of having some hard-coded mapping in the handler to understand that `scanner-v4-defs-4.5.zip` is the same as `v1`. So the mapping would be first:

```
k8s-istio.zip
scanner-defs.zip
scanner-v4-defs-4.5.zip
```

And then when `v2` arrives:

```
k8s-istio.zip
scanner-defs.zip
scanner-v4-defs-4.5.zip
scanner-v4-defs-v2.zip
```

But that's hacky.

Reply:
> Bundle would be approaching 1GB. This seems wasteful and cost prohibitive for an ACS install that would only ever need one version. We had previously communicated that there was no guarantee post 4.4 we would be adding additional versions to this bundle.

Is it? I was assuming ~1GiB was fine, for the benefit of simplifying the customer's workflow (use latest, don't worry about version), and the implementation (no need to expose the <release> -> <bundle version> mapping to places (roxctl, central handler, etc). But, if that's not the case, let's update this to deprecate the latest offline bundle, and move forward with version offline bundles (which I think is what we had agreed to do before).

Reply:
> And once 4.5 is out, we can remove it from the bundle.

If current user behaviors are an indicator - we may not be able to remove 4.5 until possibly 'years' after its released / out of support (more so when factoring in exceptions)

> Is it? I was assuming ~1GiB was fine

The net pipe to get data from the 'web' in/out of a heavily isolated environment may be intentionally limited and heavily inspected/monitored. In past experience I have faced issues transmitting smaller files.

> for the benefit of simplifying the customer's workflow (use latest, don't worry about version), and the implementation

That was the intention behind `roxctl download scanner-db` - don't need to work about version - always get the 'right file' (in theory), that would allow us to make things as complicated or as simple as behind the scenes and the user will have a single experience.

> (no need to expose the -> mapping to places (roxctl, central handler, etc

Don't know the details behind the mapping file and how its stored/structured, but we will know for a 'version' of ACS the value of `VULNERABILITY_VERSION` will be static right?

We shouldn't need to expose the mapping, `roxctl download scanner-db` will get the 'version' to download from Central, which in turn can get the version from scanner (or when we build Central also include version so it can be returned via a simple API call).  

Having that version available also enables doing validation.

Reply:
> If current user behaviors are an indicator - we may not be able to remove 4.5 until possibly 'years' after its released / out of support (more so when factoring in exceptions)

Sure, I am considering that when I said "once 4.5 is out". In other words, I am willing to keep 4.5 for some time.

> The net pipe to get data from the 'web' in/out of a heavily isolated environment may be intentionally limited and heavily inspected/monitored. In past experience I have faced issues transmitting smaller files.

Do you have some numbers of what would be painful or even problematic? I am curious to understand how moving from ~400MiB to ~600MiB would be an issue.

> That was the intention behind roxctl download scanner-db - don't need to worry about version - always get the 'right file' (in theory), that would allow us to make things as complicated or as simple as we want behind the scenes - the user will have a single experience.

I am aware of `roxctl`'s support for pulling the bundle. But I still think that customers would prefer not migrating to `roxctl` if needed. I agree that if customers all use `roxctl` things would be better for us, but better from them might be "keep using a single offline bundle".

> We shouldn't need to expose the mapping, roxctl download scanner-db will get the 'version' to download from Central, 

I am also aware of the potential implementation for this, and how the one you suggested is possible. I am sorry, I should have left a "consequence" entry for that signaling I am not proposing using latest because I don't know how it could be done with offline bundle versions -- I am actually thinking it might not be worth doing it.

I think right now I am considering the size impact more than anything -- if the size increase in the latest bundle is indeed painful to the point it make the flow worse for customers compared to migrating to `roxctl` in their scripts, then I would change my opinion on moving forward with the versioned offline bundles.

Reply:
> Do you have some numbers of what would be painful or even problematic? I am curious to understand how moving from ~400MiB to ~600MiB would be an issue.

No, those would differ per user, environment, etc.